### PR TITLE
libsyntax/parse: fix missing kind error reporting

### LINKED ISSUE #41161: "Error constructed but not emitted" ICE on nightly parsing wonky impl
rustc 1.18.0-nightly (53f4bc311 2017-04-07)
stable/beta/nightly from a week ago seem unaffected.
```
$ cat > bug.rs
impl X{Y}
$ RUST_BACKTRACE=full rustc bug.rs 
error: internal compiler error: Error constructed but not emitted
 
error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: run with `RUST_BACKTRACE=1` for a backtrace

thread 'rustc' panicked at 'explicit panic', src/librustc_errors/diagnostic_builder.rs:192
stack backtrace:
   0:        0x10fd12e83 - std::sys::imp::backtrace::tracing::imp::unwind_backtrace::hcc709dd93d82a051
   1:        0x10fd1fafd - std::panicking::default_hook::{{closure}}::he392668e67778d9e 
   2:        0x10fd1f644 - std::panicking::default_hook::h2ade4aa81d4f8204
   3:        0x10fd22277 - std::panicking::rust_panic_with_hook::h4573918fc9f80a34
   4:        0x10d2aa534 - std::panicking::begin_panic::h2a5fc248b95a3db2
   5:        0x10d2b143b - <rustc_errors::diagnostic_builder::DiagnosticBuilder<'a> as core::ops::Drop>::drop::h12de2c56938d20bb
   6:        0x10cdd289b - syntax::parse::parser::Parser::parse_impl_item::h3ddc1b71a9135797
   7:        0x10cdd50c1 - syntax::parse::parser::Parser::parse_item_impl::h284e44c48881ab63
   8:        0x10cde1356 - syntax::parse::parser::Parser::parse_item_::h44415882723eb725
   9:        0x10cdeabd8 - syntax::parse::parser::Parser::parse_item::h923de9a5bd86a0a8
  10:        0x10cdd8d5e - syntax::parse::parser::Parser::parse_mod_items::hf3c521b150993a7f
  11:        0x10cdeb7dd - syntax::parse::parser::Parser::parse_crate_mod::h19282d20641c273e
  12:        0x10cdfb19e - syntax::parse::parse_crate_from_file::hdab1619deb4b8e47
  13:        0x10b02655c - rustc_driver::driver::phase_1_parse_input::{{closure}}::hadf9f24a52d977ac
  14:        0x10b0255c5 - rustc_driver::driver::phase_1_parse_input::h4228cea37031992a
  15:        0x10b018554 - rustc_driver::driver::compile_input::h163f6d3afe612689
  16:        0x10b076046 - rustc_driver::run_compiler::h870480169c099057
  17:        0x10af49fa8 - std::panicking::try::do_call::h122fb7737645584e
  18:        0x10fd2508a - __rust_maybe_catch_panic
  19:        0x10af907ef - <F as alloc::boxed::FnBox<A>>::call_box::hdbb485c2e5737f1e
  20:        0x10fd1eae4 - std::sys::imp::thread::Thread::new::thread_start::hba7593f2e1f1beb1
  21:     0x7fff884f899c - _pthread_body
  22:     0x7fff884f8919 - _pthread_start
```

Reduced from `impl Iterator for Foo { Item = Bar; ... }` where I forgot the `type`.

### LINKED ISSUE #41239: Nightly panics `<rustc_errors::diagnostic_builder::DiagnosticBuilder<'a> as core::ops::Drop>::drop`
```
note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: run with `RUST_BACKTRACE=1` for a backtrace

thread 'rustc' panicked at 'explicit panic', /checkout/src/librustc_errors/diagnostic_builder.rs:192
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: <rustc_errors::diagnostic_builder::DiagnosticBuilder<'a> as core::ops::Drop>::drop
   1: syntax::parse::parser::Parser::parse_impl_item
   2: syntax::parse::parser::Parser::parse_item_impl
   3: syntax::parse::parser::Parser::parse_item_
   4: syntax::parse::parser::Parser::parse_item
   5: syntax::parse::parser::Parser::parse_mod_items
   6: syntax::parse::parser::Parser::parse_crate_mod
   7: syntax::parse::parse_crate_from_file
   8: rustc_driver::driver::phase_1_parse_input::{{closure}}
   9: rustc_driver::driver::phase_1_parse_input
  10: rustc_driver::driver::compile_input
  11: rustc_driver::run_compiler
  12: std::panicking::try::do_call
  13: __rust_maybe_catch_panic
             at /checkout/src/libpanic_unwind/lib.rs:98
  14: <F as alloc::boxed::FnBox<A>>::call_box
  15: std::sys::imp::thread::Thread::new::thread_start
             at /checkout/src/liballoc/boxed.rs:650
             at /checkout/src/libstd/sys_common/thread.rs:21
             at /checkout/src/libstd/sys/unix/thread.rs:84
  16: start_thread
  17: clone

error: Could not compile `slog-atomic`.

To learn more, run the command again with --verbose.
```

while compiling `slog-atomic = "2.0.0-3.1"` with the following change (I was just testing errors in Vim and introduced random error):

```
diff --git a/lib.rs b/lib.rs
index 6ac461e..ac47c1b 100644
--- a/lib.rs
+++ b/lib.rs
@@ -59,7 +59,7 @@ impl<O, E> AtomicSwitchCtrl<O, E> {
                 -> Arc<Box<SendSyncRefUnwindSafeDrain<Ok = O, Err = E>>> {
         self.0.set(drain)
     }
-
+s
     /// Get a `AtomicSwitch` drain controlled by this `AtomicSwitchCtrl`
     pub fn drain(&self) -> AtomicSwitch<O, E> {
         AtomicSwitch(self.0.clone())
```

### PR DESCRIPTION
**Pull request by arielb1**
Fixes #41161.
Fixes #41239.

### CODE REVIEW

**Comment by arielb1** [comment]
r? @petrochenkov 
cc @jonathandturner for help with the error message
cc @estebank

File: src/libsyntax/parse/parser.rs [hunk 1]
```diff
@@ -4739,34 +4739,28 @@ impl<'a> Parser<'a> {
...
-                    err.cancel();
-                    bang_err.cancel();
-                    //     pub  path(
-                    //        ^^ `sp` below will point to this
-                    let sp = prev_span.between(self.prev_span);
-                    err = self.diagnostic()
-                        .struct_span_err(sp, "missing `fn` for method declaration");
-                    err.span_label(sp, &"missing `fn`");
-                }
+            if !self.eat(&token::Not) {
```
Comment (petrochenkov) [inline src/libsyntax/parse/parser.rs]: I'd also add `path.segments.len() == 1` to the condition for mistyped impl item declaration.

As for recovery, if the current token is `(` then `err` can be emitted immediately as non-fatal and we can parse remaining tokens as a method.
If the current token is not `:` then remaining tokens can be parsed as an associated type (statistically more likely), but there's a chance for confusing error if it's actually an associated const, and this probably doesn't matter as much as methods.

Comment (arielb1) [inline src/libsyntax/parse/parser.rs]: What to do if `path.segments.len() != 1`?

File: src/test/ui/did_you_mean/issue-41161.err [hunk 2]
```diff
@@ -0,0 +1,10 @@
+error: missing `fn`, `type`, or `const` for impl-item declaration
```
Comment (estebank) [inline src/test/ui/did_you_mean/issue-41161.err]: This file shouldn't be in the PR.

File: src/libsyntax/parse/parser.rs (hunk 1 above)
Comment (petrochenkov) [inline src/libsyntax/parse/parser.rs]: >What to do if `path.segments.len() != 1`?

Report missing `!` and continue parsing as a macro, most likely.

**Comment by arielb1** [comment]
Now with error recovery.

**Pull request merged**